### PR TITLE
Initial release

### PR DESCRIPTION
**Pull request by twolfson**
This PR is the initial release of the Underdog.io Engineering blog. It does not include any of our own articles yet as I would like to facilitate that in its own PR. Instead, this focuses on `wintersmith` and repository setup.

http://engineering.underdog.io/

In this PR:
- Added `wintersmith` and initial template
- Added release and "test" scripts
- Added `npm` commands to run different `wintersmith` actions/our builds
- Updated CSS to be more inline with http://underdog.io/
  - Notable changes:
    - Removed majority of italics/one-off fonts/dashed borders
    - Moved to Lato as default body and `#333` as body color
    - Removed `about` section on main page
- Added documentation
- Removed majority of default articles
- Added GitHub over email as the link for an author

selection_359

selection_358

selection_360

/cc @brettlangdon

### CODE REVIEW

File: bin/release.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+set -e
+set -x
+
+# Build our files
+npm run build
+
+# Generate a timestamp for our release
+version="$(date --utc +%Y%m%d.%H%M%S.%N)"
+
+# Version it via `npm` (bumps package.json and adds git commit)
```
Comment (brettlangdon) [inline bin/release.sh]: `git commit and tag`

File: config.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    "moment": "moment",
+    "_": "underscore",
+    "typogr": "typogr"
+  },
+  "jade": {
+    "pretty": true
+  },
+  "markdown": {
+    "smartLists": true,
+    "smartypants": true
```
Comment (brettlangdon) [inline config.json:21]: `"gfm": true`

File: templates/archive.jade [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+						- var monthsK = _.chain(months).keys().reverse().value()
+						- for(month in monthsK){
+							- var monthHeight = lineHeight * months[monthsK[month]].length
+							li
+								span.month-label(style='line-height:'+monthHeight+'em')=month_names[monthsK[month]]
+								ul(style='margin-left:7em')
+									each item in months[monthsK[month]]
+										li(style='height:'+ lineHeight + 'em;line-height:'+ lineHeight + 'em;')
+											a(href=item.url)=item.title
+						- }
```
Comment (brettlangdon) [inline templates/archive.jade]: whoah, why is the default formatting on this file funky (not your fault, messed up on my blog too).

File: templates/author.jade [hunk 4]
```diff
@@ -0,0 +1,8 @@
+
+mixin author(authorName)
+  - var author = contents.authors[authorName + '.json']
+  span.author
+    if author
+      a(href="https://github.com/#{author.metadata.github}")= author.metadata.name
```
Comment (brettlangdon) [inline templates/author.jade]: `target="_blank"`?

File: templates/layout.jade [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+      div.content-wrap
+        block header
+          div.logo
+            h1
+              a(href=locals.url)= locals.name
+            p.description= locals.description
+    div#content
+      div.content-wrap
+        block content
+          h2 Welcome to zombocom!
```
Comment (brettlangdon) [inline templates/layout.jade]: http://www.zombo.com/

image

**Comment by brettlangdon** [comment]
looks good, few small comments, but nothing too serious

**Comment by twolfson** [comment]
Hmm, iojs is being a troublesome build. I might reduce Travis CI to `node@0.10` only =_=

File: templates/layout.jade (hunk 5 above)
Comment (twolfson) [inline templates/layout.jade]: Buuut we replace the content in all places. Why didn't they leave this blank or with a comment =_=

File: templates/author.jade (hunk 4 above)
Comment (twolfson) [inline templates/author.jade]: Yea, we could. I kind of think a normal link would make sense but that's me.

File: templates/archive.jade (hunk 3 above)
Comment (twolfson) [inline templates/archive.jade]: I don't think the `wintersmith` repo does code consistency too much =(

File: config.json (hunk 2 above)
Comment (twolfson) [inline config.json:21]: What does this enable that we never had before? It looks like code blocks are getting parsed and highlighted.

File: bin/release.sh (hunk 1 above)
Comment (twolfson) [inline bin/release.sh]: :+1:

File: templates/article.jade [hunk 6]
```diff
@@ -0,0 +1,25 @@
+
+extends layout
+
+block append vars
+  - bodyclass = 'article-detail'
+
+block prepend title
+  | #{ page.title + " - "}
```
Comment (twolfson) [inline templates/article.jade]: Oooh, I dig this for handling trailing whitespace

File: templates/article.jade [hunk 7]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (twolfson) [inline templates/article.jade]: We should remove this leading whitespace

File: templates/article.jade [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+block content
+  article.article
+    section.content!= typogr(page.html).typogrify()
+
+block prepend footer
+  div.nav
+    a(href=contents.index.url) « Full blog
+
+
```
Comment (twolfson) [inline templates/article.jade]: We should remove this trailing whitespace

File: templates/author.jade [hunk 9]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (twolfson) [inline templates/author.jade]: We should remove this leading whitespace

File: templates/archive.jade [hunk 10]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (twolfson) [inline templates/archive.jade]: Leading whitespace

File: plugins/paginator.coffee [hunk 11]
```diff
@@ -0,0 +1,91 @@
+
```
Comment (twolfson) [inline plugins/paginator.coffee]: Leading whitespace

File: contents/css/main.css [hunk 12]
```diff
@@ -0,0 +1,419 @@
+
```
Comment (twolfson) [inline contents/css/main.css]: Leading whitespace

File: templates/index.jade [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  div.nav
+    if prevPage
+      a(href=prevPage.url) « Newer
+    else
+      a(href='/archive.html') « Archives
+    if nextPage
+      a(href=nextPage.url) Next page »
+
+
+
```
Comment (twolfson) [inline templates/index.jade]: Traaailing whitespace

File: contents/css/main.css [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+    -webkit-column-count: 1;
+    -moz-column-count: 1;
+    -ms-column-count: 1;
+    column-count: 1;
+  }
+  footer .about p {
+    margin-bottom: 1em;
+  }
+}
+
```
Comment (twolfson) [inline contents/css/main.css]: Trailing whitespace

File: templates/author.jade [hunk 15]
```diff
@@ -0,0 +1,8 @@
...
+mixin author(authorName)
+  - var author = contents.authors[authorName + '.json']
+  span.author
+    if author
+      a(href="https://github.com/#{author.metadata.github}")= author.metadata.name
+    else
+      =authorName
```
Comment (twolfson) [inline templates/author.jade]: No space between var and equals

File: templates/article.jade (hunk 6 above)
Comment (twolfson) [inline templates/article.jade]: Inconsistent spacing between `#{` and other instances. Prob go for `#{page.title + " - "}`

File: templates/article.jade [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+  - bodyclass = 'article-detail'
+
+block prepend title
+  | #{ page.title + " - "}
+
+block header
+  include author
+  h1= page.title
+  p.author
+    | #{ "Written by " }
```
Comment (twolfson) [inline templates/article.jade]: Fix up spacing for interpolation `= "Written by "`

Comment (twolfson) [inline templates/article.jade]: Or keep it consistent with `| #{"Written by"}` but this seems like a lot of excess

File: templates/archive.jade [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+extends layout
+//- this logic should be moved to a view at some point
+
+block content
+	- var lineHeight = 2.2;
+	- var archives = _.chain(env.helpers.getArticles(contents)).groupBy(function(item) {
```
Comment (twolfson) [inline templates/archive.jade]: Fix function spacing to be consistent with our code style

File: templates/archive.jade [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+	-	 return item.date.getFullYear()
+	- }).value()
+	- for (var archive in archives) {
+	-	archives[archive] = _.groupBy(archives[archive], function(item){return item.date.getMonth();})
```
Comment (twolfson) [inline templates/archive.jade]: Fix function spacing to be consistent with our coding style

File: templates/archive.jade [hunk 19]
```diff
@@ -0,0 +1,34 @@
+
+extends layout
+//- this logic should be moved to a view at some point
+
+block content
+	- var lineHeight = 2.2;
+	- var archives = _.chain(env.helpers.getArticles(contents)).groupBy(function(item) {
+	-	 return item.date.getFullYear()
```
Comment (twolfson) [inline templates/archive.jade]: Missing semicolon

File: templates/archive.jade [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+	- }).value()
+	- for (var archive in archives) {
+	-	archives[archive] = _.groupBy(archives[archive], function(item){return item.date.getMonth();})
+	- }
+	- var month_names = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
+	section.archive
+		h2 Archive
+		ul
+			- var yearsK = _.chain(archives).keys().reverse().value()
+			- for(var year in yearsK)
```
Comment (twolfson) [inline templates/archive.jade]: Missing a curly brace for this loop

Comment (twolfson) [inline templates/archive.jade]: Better yet, why not use a `jade` loop?

Comment (twolfson) [inline templates/archive.jade]:
``` jade
for year in yearsK
```

File: templates/archive.jade [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+				- months = archives[yearsK[year]]
+				- var yearHeight = lineHeight * _.reduce(months, function(memo,month) { return memo + month.length; }, 0);
```
Comment (twolfson) [inline templates/archive.jade]: Function spacing is inconsistent

File: templates/archive.jade [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+				li
+					span.year-label(style='line-height:' + yearHeight+'em')=yearsK[year]
```
Comment (twolfson) [inline templates/archive.jade]: Use double quotes for Jade. Spacing inconsistent for `+'s`. Need space between `=` and `yearsK`

File: templates/archive.jade [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+					ul(style='margin-left:4em')
```
Comment (twolfson) [inline templates/archive.jade]: Use double quotes for Jade. Can we abstract the inline style?

File: templates/archive.jade [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+						- var monthsK = _.chain(months).keys().reverse().value()
+						- for(month in monthsK){
```
Comment (twolfson) [inline templates/archive.jade]: Use Jade loop instead

File: templates/archive.jade [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+							- var monthHeight = lineHeight * months[monthsK[month]].length
+							li
+								span.month-label(style='line-height:'+monthHeight+'em')=month_names[monthsK[month]]
```
Comment (twolfson) [inline templates/archive.jade]: Double quotes and `+` spacing

Comment (twolfson) [inline templates/archive.jade]: I think Jade support using an object for style so we can make this interpolation:

``` jade
span.month-label(style={"line-height": "#{monthHeight}em"})
```

File: templates/archive.jade [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+								ul(style='margin-left:7em')
```
Comment (twolfson) [inline templates/archive.jade]: Double quotes and try to abstract style

File: templates/archive.jade [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+									each item in months[monthsK[month]]
```
Comment (twolfson) [inline templates/archive.jade]: Hooray, Jade loop =D

File: templates/archive.jade [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+										li(style='height:'+ lineHeight + 'em;line-height:'+ lineHeight + 'em;')
```
Comment (twolfson) [inline templates/archive.jade]: Double quotes, `+` spacing, and maybe Jade abstraction for `style`

File: templates/archive.jade [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+											a(href=item.url)=item.title
+						- }
+
```
Comment (twolfson) [inline templates/archive.jade]: Trailing whitespace

File: plugins/paginator.coffee [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+    articles = contents[options.articles]._.directories.map (item) -> item.index
+    # skip articles that does not have a template associated
+    articles = articles.filter (item) -> item.template isnt 'none'
+    # sort article by date
+    articles.sort (a, b) -> b.date - a.date
+    return articles
+
+  class PaginatorPage extends env.plugins.Page
+    ### A page has a number and a list of articles ###
+
```
Comment (twolfson) [inline plugins/paginator.coffee]: Leading whitespace

File: plugins/paginator.coffee [hunk 31]
```diff
@@ -0,0 +1,91 @@
...
+      # extend the template context with the enviroment locals
+      env.utils.extend ctx, locals
+
+      # finally render the template
+      template.render ctx, callback
+
+  # register a generator, 'paginator' here is the content group generated content will belong to
+  # i.e. contents._.paginator
+  env.registerGenerator 'paginator', (contents, callback) ->
+
```
Comment (twolfson) [inline plugins/paginator.coffee]: Leading whitespace

File: templates/feed.jade [hunk 32]
```diff
@@ -0,0 +1,25 @@
+doctype xml
+rss(version='2.0',
+    xmlns:content='http://purl.org/rss/1.0/modules/content/',
+    xmlns:wfw='http://wellformedweb.org/CommentAPI/',
+    xmlns:dc='http://purl.org/dc/elements/1.1/'
+    xmlns:atom='http://www.w3.org/2005/Atom')
```
Comment (twolfson) [inline templates/feed.jade]: Use double quotes for Jade

File: templates/feed.jade [hunk 33]
```diff
@@ -0,0 +1,25 @@
...
+  channel
+    - var articles = env.helpers.getArticles(contents);
+    title= locals.name
+    atom:link(href=locals.url + '/feed.xml', rel='self', type='application/rss+xml')
```
Comment (twolfson) [inline templates/feed.jade]: Double quotes and string interpolation

File: templates/index.jade [hunk 34]
```diff
@@ -0,0 +1,29 @@
...
+        if article.hasMore
+          p.more
+            a(href=article.url) more
+
+block prepend footer
+  div.nav
+    if prevPage
+      a(href=prevPage.url) « Newer
+    else
+      a(href='/archive.html') « Archives
```
Comment (twolfson) [inline templates/index.jade]: Double quotes

File: templates/layout.jade [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+    block head
+      meta(charset='utf-8')
+      meta(http-equiv='X-UA-Compatible', content='IE=edge,chrome=1')
+      meta(name='viewport', content='width=device-width')
+      title
+        block title
+          = locals.name
+      link(rel='alternate', href=locals.url+'/feed.xml', type='application/rss+xml', title=locals.description)
+      link(rel='stylesheet', href='http://fonts.googleapis.com/css?family=Lato:400,700|Anonymous+Pro:400,700,400italic,700italic')
+      link(rel='stylesheet', href=contents.css['main.css'].url)
```
Comment (twolfson) [inline templates/layout.jade]: Double quotes

File: templates/layout.jade [hunk 36]
```diff
@@ -0,0 +1,34 @@
...
+            p.description= locals.description
+    div#content
+      div.content-wrap
+        block content
+          h2 Welcome to zombocom!
+    footer
+      div.content-wrap
+        block footer
+          section.copy.muted
+            p &copy; #{ new Date().getFullYear() } #{ locals.owner } &mdash; powered by&nbsp;
```
Comment (twolfson) [inline templates/layout.jade]: Remove spacing around `#{` and `}`

File: templates/layout.jade [hunk 37]
```diff
@@ -0,0 +1,34 @@
...
+              a(href='https://github.com/jnordberg/wintersmith') Wintersmith
```
Comment (twolfson) [inline templates/layout.jade]: Double quotes

**Comment by twolfson** [comment]
I rushed through this on the first run. Left of a lot of comments to myself about making the code style consistent. Maybe we should open it as a PR on the `wintersmith` repo?

**Comment by brettlangdon** [comment]
So many comments!

Yeah, that is what I was thinking too.

**Comment by twolfson** [comment]
I was overzealous with some of my comments last time. Going to delete some now.

File: config.json (hunk 2 above)
Comment (twolfson) [inline config.json:21]: Did some research. `wintersmith` uses `marked`. The version it's pointing to has `gfm` enabled by default:

https://github.com/jnordberg/wintersmith/blob/2.2.0/src/plugins/markdown.coffee#L4

https://github.com/jnordberg/wintersmith/blob/2.2.0/package.json#L35

https://github.com/chjj/marked/tree/v0.3.3#gfm

**Comment by brettlangdon** [comment]
:+1:

**Pull request merged**